impl/src/main/java/org/jboss/weld/config/ConfigurationKey.java

```diff
@@ -271,6 +274,53 @@
      */
     ALLOW_OPTIMIZED_CLEANUP("org.jboss.weld.bootstrap.allowOptimizedCleanup", false, true),
 
+    /**
+     * A regular expression. If {@link #ALLOW_OPTIMIZED_CLEANUP} is set to true this property can be used to extend the set of beans which should never be
+     * considered <strong>unused</strong>. {@link Bean#getBeanClass()} is used to match the pattern.
+     *
+     * <p>
+     * By default, all beans are excluded. If undefined (an empty string), no beans are excluded.
+     * </p>
+     *
+     * An unused bean:
+     * <ul>
+     * <li>is not excluded by this property or {@link #UNUSED_BEANS_EXCLUDE_ANNOTATION}</li>
+     * <li>is not a built-in bean, session bean, extension, interceptor or decorator,</li>
+     * <li>does not have a name</li>
+     * <li>does not declare an observer</li>
+     * <li>is not eligible for injection to any injection point</li>
+     * <li>does not declare a producer which is eligible for injection to any injection point</li>
+     * <li>is not eligible for injection into any {@link Instance} injection point</li>
+     * </ul>
+     *
+     * @see ConfigurationKey#UNUSED_BEANS_EXCLUDE_ANNOTATION
+     */
+    UNUSED_BEANS_EXCLUDE_TYPE("org.jboss.weld.bootstrap.unusedBeans.excludeType", ".*"),
```

Comment:
So, if I get this correctly, we are basically going to be eager (if the cleanup is enabled) and these two options are safety latches which can be utilized to avoid cleaning some beans. And the reason is that we cannot detect, for instance, dynamic resolution.

Reply:
More or less. By default, all types are excluded, i.e. the feature is de facto disabled. If you set this property to empty string (or reg exp which does not match anything) then no type is excluded. You can also specify a reg exp e.g. to exclude classes from your app. `UNUSED_BEANS_EXCLUDE_ANNOTATION` is an additional "safety latch" to exclude types considering annotations only (e.g. by default any JAX-RS annotation -> any JAX-RS resource).